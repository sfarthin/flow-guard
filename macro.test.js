const write = require('fs').writeFileSync;
const babel = require('@babel/core');
const vm = require('vm');
const decoders = require('decoders');

const filename = `${__dirname}/__testing__.js`;

const prefix = `// @flow\nconst guard = require('./macro');\n`;

const t = (guardCode, guardName = 'guard') => {
  const src = `${guardCode}`;
  write(filename, src);
  /**
   * Make Snapshot
   */
  const code = babel.transformSync(src, {
    filename,
    plugins: [
      '@babel/plugin-transform-modules-commonjs',
      '@babel/plugin-syntax-flow',
      'macros',
    ],
  }).code;
  expect(code).toMatchSnapshot();

  /**
   * Make certain its executable
   */
  const script = new vm.Script(code);
  const sandbox = { require: () => decoders };
  vm.createContext(sandbox);
  script.runInContext(sandbox);
};

const tests = [
  ['string', '"foo"'],
  ['number', '14'],
  ['boolean', 'true'],
  ['String', '"foo"'],
  ['Number', '14'],
  ['Boolean', 'true'],
  ['?string', 'null'],
  ['?string | boolean | Number', '3'],
  ['string | null', 'null'],
  ['{ a: { b: string }, c: number }', "{ a: { b: 'b' }, c: 2 }"],
  ['"foo"', "'foo'"],
  ['5', '5'],
  ['true', 'true'],
];

it('can create inline checks', () => {
  // Test all the inline types (pulls from Babel AST).
  tests.forEach(([a, b]) => {
    t(`${prefix}guard<${a}>(${b});\n`);
  });
});

it('can create aliased checks', () => {
  // Test all the aliased types (infered from Flow AST).
  tests.forEach(([a, b]) => {
    t(`${prefix}type Foo = ${a};
guard<Foo>(${b});\n`);
  });
});

it('can create inline checks with shadowed variables', () => {
  tests.forEach(([a, b]) => {
    t(`// @flow
      const g = require('./macro');

      function foo() {
        const guard = 'foo';
        const string = 'foo';
        const boolean = 'foo';
        const object = 'foo';

        g<${a}>(${b});
      }
      foo();
    `);
  });
});

it('can create aliased checks with shadowed variables', () => {
  tests.forEach(([a, b]) => {
    t(`// @flow
        type Foo = ${a};
        const g = require('./macro');
        const guard = 'foo';
        const string = 'foo';
        const boolean = 'foo';
        const object = 'foo';

        function foo() {
          const gg = '4';
          {
            const ggg = '4';
            g<Foo>(${b});
          }

        }
        foo();
      `);
  });
});

it('works with decoders', () => {
  tests.forEach(([a, b]) => {
    t(`// @flow
      type Foo = ${a};
      const guard = require('decoders').guard;
      const createDecoder = require('./decoder.macro');
      const string = 'foo';
      const boolean = 'foo';
      const object = 'foo';

      function foo() {
        const gg = '4';
        {
          const ggg = '4';
          const decoder = createDecoder<Foo>();
          guard(decoder)(${b})
        }

      }
      foo();
    `);
  });
});
