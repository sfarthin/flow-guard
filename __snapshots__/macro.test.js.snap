// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create aliased checks 1`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.string)(\\"foo\\");"
`;

exports[`can create aliased checks 2`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.number)(14);"
`;

exports[`can create aliased checks 3`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.boolean)(true);"
`;

exports[`can create aliased checks 4`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.string)(\\"foo\\");"
`;

exports[`can create aliased checks 5`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.number)(14);"
`;

exports[`can create aliased checks 6`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.boolean)(true);"
`;

exports[`can create aliased checks 7`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.either3)(_decoders.undefined_, _decoders.null_, _decoders.string))(null);"
`;

exports[`can create aliased checks 8`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.either5)(_decoders.undefined_, _decoders.null_, _decoders.string, _decoders.boolean, _decoders.number))(3);"
`;

exports[`can create aliased checks 9`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.either)(_decoders.string, _decoders.null_))(null);"
`;

exports[`can create aliased checks 10`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.object)({
  a: (0, _decoders.object)({
    b: _decoders.string
  }),
  c: _decoders.number
}))({
  a: {
    b: 'b'
  },
  c: 2
});"
`;

exports[`can create aliased checks 11`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(\\"foo\\"))('foo');"
`;

exports[`can create aliased checks 12`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(5))(5);"
`;

exports[`can create aliased checks 13`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(true))(true);"
`;

exports[`can create aliased checks with shadowed variables 1`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.string)(\\"foo\\");
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 2`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.number)(14);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 3`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.boolean)(true);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 4`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.string)(\\"foo\\");
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 5`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.number)(14);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 6`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)(_decoders.boolean)(true);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 7`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.either3)(_decoders.undefined_, _decoders.null_, _decoders.string))(null);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 8`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.either5)(_decoders.undefined_, _decoders.null_, _decoders.string, _decoders.boolean, _decoders.number))(3);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 9`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.either)(_decoders.string, _decoders.null_))(null);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 10`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.object)({
      a: (0, _decoders.object)({
        b: _decoders.string
      }),
      c: _decoders.number
    }))({
      a: {
        b: 'b'
      },
      c: 2
    });
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 11`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.constant)(\\"foo\\"))('foo');
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 12`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.constant)(5))(5);
  }
}

foo();"
`;

exports[`can create aliased checks with shadowed variables 13`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = 'foo';
const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    (0, _decoders.guard)((0, _decoders.constant)(true))(true);
  }
}

foo();"
`;

exports[`can create inline checks 1`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.string)(\\"foo\\");"
`;

exports[`can create inline checks 2`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.number)(14);"
`;

exports[`can create inline checks 3`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.boolean)(true);"
`;

exports[`can create inline checks 4`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.string)(\\"foo\\");"
`;

exports[`can create inline checks 5`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.number)(14);"
`;

exports[`can create inline checks 6`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)(_decoders.boolean)(true);"
`;

exports[`can create inline checks 7`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.nullable)(_decoders.string))(null);"
`;

exports[`can create inline checks 8`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.either3)((0, _decoders.nullable)(_decoders.string), _decoders.boolean, _decoders.number))(3);"
`;

exports[`can create inline checks 9`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.either)(_decoders.string, _decoders.null_))(null);"
`;

exports[`can create inline checks 10`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.object)({
  a: (0, _decoders.object)({
    b: _decoders.string
  }),
  c: _decoders.number
}))({
  a: {
    b: 'b'
  },
  c: 2
});"
`;

exports[`can create inline checks 11`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(\\"foo\\"))('foo');"
`;

exports[`can create inline checks 12`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(5))(5);"
`;

exports[`can create inline checks 13`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

(0, _decoders.guard)((0, _decoders.constant)(true))(true);"
`;

exports[`can create inline checks with shadowed variables 1`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.string)(\\"foo\\");
}

foo();"
`;

exports[`can create inline checks with shadowed variables 2`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.number)(14);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 3`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.boolean)(true);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 4`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.string)(\\"foo\\");
}

foo();"
`;

exports[`can create inline checks with shadowed variables 5`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.number)(14);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 6`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)(_decoders.boolean)(true);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 7`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.nullable)(_decoders.string))(null);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 8`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.either3)((0, _decoders.nullable)(_decoders.string), _decoders.boolean, _decoders.number))(3);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 9`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.either)(_decoders.string, _decoders.null_))(null);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 10`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.object)({
    a: (0, _decoders.object)({
      b: _decoders.string
    }),
    c: _decoders.number
  }))({
    a: {
      b: 'b'
    },
    c: 2
  });
}

foo();"
`;

exports[`can create inline checks with shadowed variables 11`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.constant)(\\"foo\\"))('foo');
}

foo();"
`;

exports[`can create inline checks with shadowed variables 12`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.constant)(5))(5);
}

foo();"
`;

exports[`can create inline checks with shadowed variables 13`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

function foo() {
  const guard = 'foo';
  const string = 'foo';
  const boolean = 'foo';
  const object = 'foo';
  (0, _decoders.guard)((0, _decoders.constant)(true))(true);
}

foo();"
`;

exports[`works with decoders 1`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.string;
    guard(decoder)(\\"foo\\");
  }
}

foo();"
`;

exports[`works with decoders 2`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.number;
    guard(decoder)(14);
  }
}

foo();"
`;

exports[`works with decoders 3`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.boolean;
    guard(decoder)(true);
  }
}

foo();"
`;

exports[`works with decoders 4`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.string;
    guard(decoder)(\\"foo\\");
  }
}

foo();"
`;

exports[`works with decoders 5`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.number;
    guard(decoder)(14);
  }
}

foo();"
`;

exports[`works with decoders 6`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = _decoders.boolean;
    guard(decoder)(true);
  }
}

foo();"
`;

exports[`works with decoders 7`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.either3)(_decoders.undefined_, _decoders.null_, _decoders.string);
    guard(decoder)(null);
  }
}

foo();"
`;

exports[`works with decoders 8`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.either5)(_decoders.undefined_, _decoders.null_, _decoders.string, _decoders.boolean, _decoders.number);
    guard(decoder)(3);
  }
}

foo();"
`;

exports[`works with decoders 9`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.either)(_decoders.string, _decoders.null_);
    guard(decoder)(null);
  }
}

foo();"
`;

exports[`works with decoders 10`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.object)({
      a: (0, _decoders.object)({
        b: _decoders.string
      }),
      c: _decoders.number
    });
    guard(decoder)({
      a: {
        b: 'b'
      },
      c: 2
    });
  }
}

foo();"
`;

exports[`works with decoders 11`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.constant)(\\"foo\\");
    guard(decoder)('foo');
  }
}

foo();"
`;

exports[`works with decoders 12`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.constant)(5);
    guard(decoder)(5);
  }
}

foo();"
`;

exports[`works with decoders 13`] = `
"\\"use strict\\";

var _decoders = require(\\"decoders\\");

const guard = require('decoders').guard;

const string = 'foo';
const boolean = 'foo';
const object = 'foo';

function foo() {
  const gg = '4';
  {
    const ggg = '4';
    const decoder = (0, _decoders.constant)(true);
    guard(decoder)(true);
  }
}

foo();"
`;
